### PR TITLE
:sparkles: Implement Wait functionality

### PR DESCRIPTION
**Pull request by janober**
First PoC for long-term wait/sleep functionality.

Is still WIP and not properly tested. Currently only works with SQLite.

Got developed on top of tag-branch to minimize the number of merge conflicts.

### CODE REVIEW

**Comment by lublak** [comment]
Just as a question: Couldn't you use the cron node instead of the wait node?
And instead of: putExecutionToSleep -> sleepTill accept a trigger node?
This would allow you to wait not only for a certain time, but also for other events in the future.

**Comment by janober** [comment]
Honestly do not understand how the Cron-Node could be used. After all is it a Trigger-Node. What would be possible is to add similar parameters to the Wait-Node as the Cron-Node contains (like to start the workflow next Thursday at 5pm again). But do not think it should be possible to retrigger the execution multiple times. If anybody wants that, this is the totally wrong functionality. After all, is it stopping and starting again a specific execution. As soon as triggering it multiple times is possible it should start new separate executions.

Just giving it a node would not be great. Each node needs very specific functionality. For the waiting for X amount of time, n8n has to keep track of that internally in an efficient way. If there are for example a few thousand executions waiting for 2 months I would like to avoid having a timer for each of them active for 2 months constantly.

Being able to restart a workflow again via a Webhook-Node (which should also be possible in the future) would for example need then the functionality to provide a totally different URL which includes not just the path but also the Execution-ID. But also that would work again totally differently underneath the hood than the Webhook-Node currently works. So would it for example not register it in the webhooks table.

**Comment by lublak** [comment]
@janober 
It's also so great for me. This idea just came to me. But I think it's probably not very efficient and a waiting period is more efficient.
Basic should actually be enough, because you usually want to delay something.

My idea came from the fact that we work a lot with files and at a certain point it can happen that a file is not yet available. So instead of using a "wait" node to check if the file is there, the idea would be to use a file watcher node.

`[some workflow start] => [file exists] => [if false] => [file watcher node] `

**Comment by janober** [comment]
Makes sense. We could also additionally implement something like that. With similar underlying code to Interval, Cron, RabbitMQ, ... nodes. So the ones that use the "trigger" function. The only problem there would be, that each waiting execution would eat quite some resources. We have the same problems with active workflows (which use exactly that trigger nodes) but because there is normally a reasonable small amount of active workflows (< 100) that never caused problems. But there could be a quite large amount of waiting executions and then it could definitely turn into a problem at some point.

Even though it is not perfect, it could still be done. It would then be on us to educate people about it to make sure that they do not get totally crazy with it and then wonder why everything is super slow and why n8n eats a lot of resources even if no execution is running.

But would choose the following order as that seems to be what most people need:
1. Wait for X amount of time
2. Wait for a webhook call
3. Everything else

**Comment by lublak** [comment]
> eat quite some resources

Good, you're right. I didn't think about the fact that triggers need so much more resources than an optimized "wait" which is needed 90% of the time anyway.
I thought you could put everything under one hat and not need multiple implementations to save work.
But there you are one step ahead of me.

**Comment by janober** [comment]
Now also PoC functionality to restart workflow via Webhook-URL got implemented. Anyway, still a lot of work before production-ready and properly usable.

**Review by BHesseldieck — COMMENTED** [review]
part (1/2)

File: packages/cli/src/WaitingWebhooks.ts [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+
+		const workflowStartNode = workflow.getNode(lastNodeExecuted);
+
+		if (workflowStartNode === null) {
+			throw new ResponseHelper.ResponseError('Could not find node to process webhook.', 404, 404);
+		}
+
+		const runExecutionData = fullExecutionData.data as IRunExecutionData;
+
+		return new Promise((resolve, reject) => {
```
[the remark this answers is not in the document]
Comment (janober) [reply packages/cli/src/WaitingWebhooks.ts:107]: I would say yes in one way or the other. the promise resolves depending on the result of that callback function.

**Comment by janober** [comment]
Got merged.

**Comment by janober** [comment]
Got released with 0.135.0

**Pull request merged**